File: packages/unity/Runtime/GameAlytics.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;

namespace GameAlytics
{
    public enum Environment
    {
        DEVELOPMENT,
        PRODUCTION
    }

    public class UserConfig
    {
        private readonly string sessionId;
        private readonly string userId;
        private readonly string anonymousId;
        
        private UserConfig(string sessionId, string userId, string anonymousId)
        {
            if (string.IsNullOrEmpty(sessionId))
            {
                throw new ArgumentException("sessionId is mandatory");
            }
            if (string.IsNullOrEmpty(userId) && string.IsNullOrEmpty(anonymousId))
            {
                throw new ArgumentException("Either userId or anonymousId must be provided");
            }
            
            this.sessionId = sessionId;
            this.userId = userId;
            this.anonymousId = anonymousId;
        }
        
        public string GetSessionId() { return sessionId; }
        public string GetUserId() { return userId; }
        public string GetAnonymousId() { return anonymousId; }
        
        public static Builder CreateBuilder()
        {
            return new Builder();
        }
        
        public class Builder
        {
            private string sessionId;
            private string userId;
            private string anonymousId;
            
            public Builder SetSessionId(string sessionId)
            {
                this.sessionId = sessionId;
                return this;
            }
            
            public Builder SetUserId(string userId)
            {
                this.userId = userId;
                return this;
            }
            
            public Builder SetAnonymous(string anonymousId)
            {
                this.anonymousId = anonymousId;
                return this;
            }
            
            public UserConfig Build()
            {
                return new UserConfig(sessionId, userId, anonymousId);
            }
        }
    }
    
    public class DeviceInfo
    {
        public readonly string platform;
        public readonly string osVersion;
        public readonly string appVersion;
        public readonly string deviceModel;
        public readonly string screenResolution;
        public readonly string deviceManufacturer;
        
        public DeviceInfo(string platform, string osVersion, string appVersion, 
                         string deviceModel, string screenResolution, string deviceManufacturer)
        {
            this.platform = platform;
            this.osVersion = osVersion;
            this.appVersion = appVersion;
            this.deviceModel = deviceModel;
            this.screenResolution = screenResolution;
            this.deviceManufacturer = deviceManufacturer;
        }
    }

    public class GameAlytics : MonoBehaviour
    {
        // Singleton instance
        private static GameAlytics instance;
        
        // Configuration
        private string apiKey;
        private Environment environment;
        private UserConfig userConfig;
        private DeviceInfo deviceInfo;
        private string baseUrl;
        private bool isInitialized = false;
        
        // Queue for events
        private readonly Queue<Dictionary<string, object>> eventQueue = new Queue<Dictionary<string, object>>();
        private bool isSending = false;
        private const int MaxQueueSize = 1000;
        private const int BatchSize = 10;

        // User Events
        public static class UserEvents
        {
            public const string CATEGORY = "user";
            public const string SESSION_START = "session_start";
            public const string SESSION_END = "session_end";
            public const string USER_LOGIN = "user_login";
            public const string USER_LOGOUT = "user_logout";
            public const string USER_REGISTER = "user_register";
        }
        
        // Gameplay Events
        public static class GameplayEvents
        {
            public const string CATEGORY = "gameplay";
            public const string LEVEL_START = "level_start";
            public const string LEVEL_END = "level_end";
            public const string LEVEL_UP = "level_up";
            public const string GAME_START = "game_start";
            public const string GAME_END = "game_end";
            public const string BOSS_FIGHT = "boss_fight";
        }
        
        // Economy Events
        public static class EconomyEvents
        {
            public const string CATEGORY = "economy";
            public const string CURRENCY_EARNED = "currency_earned";
            public const string CURRENCY_SPENT = "currency_spent";
            public const string ITEM_PURCHASED = "item_purchased";
            public const string ITEM_SOLD = "item_sold";
            public const string SHOP_VIEWED = "shop_viewed";
        }
        
        // Progression Events
        public static class ProgressionEvents
        {
            public const string CATEGORY = "progression";
            public const string TUTORIAL_COMPLETE = "tutorial_complete";
            public const string ACHIEVEMENT_UNLOCKED = "achievement_unlocked";
            public const string MILESTONE_REACHED = "milestone_reached";
            public const string QUEST_COMPLETED = "quest_completed";
        }
        
        // Ad Events
        public static class AdEvents
        {
            public const string CATEGORY = "ad";
            public const string AD_VIEWED = "ad_viewed";
            public const string AD_CLICKED = "ad_clicked";
            public const string AD_REWARDED = "ad_rewarded";
            public const string AD_FAILED = "ad_failed";
        }
        
        // IAP Events
        public static class IAPEvents
        {
            public const string CATEGORY = "iap";
            public const string PURCHASE = "purchase";
            public const string PURCHASE_FAILED = "purchase_failed";
            public const string PURCHASE_RESTORED = "purchase_restored";
            public const string SUBSCRIPTION_STARTED = "subscription_started";
            public const string SUBSCRIPTION_CANCELLED = "subscription_cancelled";
        }

        public static InitBuilder Init(string apiKey, Environment environment)
        {
            return new InitBuilder(apiKey, environment);
        }

        public static GameAlytics GetInstance()
        {
            if (instance == null)
            {
                throw new InvalidOperationException("GameAlytics must be initialized first. Call GameAlytics.Init(...).Create()");
            }
            return instance;
        }

        public SystemEventBuilder SystemEvent()
        {
            return new SystemEventBuilder();
        }

        public CustomEventBuilder CustomEvent()
        {
            return new CustomEventBuilder();
        }

        public class InitBuilder
        {
            private readonly string apiKey;
            private readonly Environment environment;
            private UserConfig userConfig;
            
            public InitBuilder(string apiKey, Environment environment)
            {
                this.apiKey = apiKey;
                this.environment = environment;
            }
            
            public InitBuilder UserConfig(UserConfig userConfig)
            {
                this.userConfig = userConfig;
                return this;
            }
            
            public GameAlytics Create()
            {
                if (string.IsNullOrEmpty(apiKey))
                {
                    throw new ArgumentException("API key is required");
                }
                if (userConfig == null)
                {
                    throw new ArgumentException("UserConfig is required");
                }
                
                if (instance == null)
                {
                    var go = new GameObject("GameAlytics");
                    instance = go.AddComponent<GameAlytics>();
                    DontDestroyOnLoad(go);
                    instance.Initialize(apiKey, environment, userConfig);
                }
                
                return instance;
            }
        }

        private void Awake()
        {
            if (instance != null && instance != this)
            {
                Destroy(gameObject);
            }
        }

        private void Initialize(string apiKey, Environment environment, UserConfig userConfig)
        {
            this.apiKey = apiKey;
            this.environment = environment;
            this.userConfig = userConfig;
            this.deviceInfo = AutoFetchDeviceInfo();
            this.baseUrl = environment == Environment.PRODUCTION 
                ? "https://client.gamealytics.click/events/collect" 
                : "https://client.dev.gamealytics.click/events/collect";
            this.isInitialized = true;
        }

        public class SystemEventBuilder
        {
            private string category;
            private string type;
            private Dictionary<string, string> properties = new Dictionary<string, string>();
            
            public SystemEventBuilder Category(Type categoryClass)
            {
                this.category = categoryClass.Name.ToLower().Replace("events", "");
                return this;
            }
            
            public SystemEventBuilder Type(string type)
            {
                this.type = type;
                return this;
            }
            
            public SystemEventBuilder SetProperties(Dictionary<string, string> properties)
            {
                this.properties = properties ?? new Dictionary<string, string>();
                return this;
            }
            
            public void Trigger()
            {
                if (instance == null)
                {
                    throw new InvalidOperationException("GameAlytics must be initialized first");
                }
                if (string.IsNullOrEmpty(category) || string.IsNullOrEmpty(type))
                {
                    throw new ArgumentException("Category and type are required");
                }
                instance.TrackEventInternal("SYSTEM", type, category, properties);
            }
        }
        
        public class CustomEventBuilder
        {
            private string category;
            private string type;
            private Dictionary<string, string> properties = new Dictionary<string, string>();
            
            public CustomEventBuilder Category(string category)
            {
                this.category = category;
                return this;
            }
            
            public CustomEventBuilder Type(string type)
            {
                this.type = type;
                return this;
            }
            
            public CustomEventBuilder SetProperties(Dictionary<string, string> properties)
            {
                this.properties = properties ?? new Dictionary<string, string>();
                return this;
            }
            
            public void Trigger()
            {
                if (instance == null)
                {
                    throw new InvalidOperationException("GameAlytics must be initialized first");
                }
                if (string.IsNullOrEmpty(category) || string.IsNullOrEmpty(type))
                {
                    throw new ArgumentException("Category and type are required");
                }
                instance.TrackEventInternal("CUSTOM", type, category, properties);
            }
        }

        private void TrackEventInternal(string eventClass, string eventType, string category, Dictionary<string, string> properties)
        {
            if (eventQueue.Count >= MaxQueueSize)
            {
                eventQueue.Dequeue();
            }
            
            var eventData = new Dictionary<string, object>
            {
                ["type"] = eventClass,
                ["value"] = eventType,
                ["category"] = category,
                ["platform"] = deviceInfo.platform,
                ["osVersion"] = deviceInfo.osVersion,
                ["deviceModel"] = deviceInfo.deviceModel,
                ["deviceManufacturer"] = deviceInfo.deviceManufacturer,
                ["appVersion"] = deviceInfo.appVersion,
                ["screenResolution"] = deviceInfo.screenResolution,
                ["userId"] = userConfig.GetUserId() ?? "",
                ["anonymousId"] = userConfig.GetAnonymousId() ?? "",
                ["sessionId"] = userConfig.GetSessionId(),
                ["timezone"] = TimeZoneInfo.Local.StandardName,
                ["localDateTime"] = DateTime.UtcNow.ToString("o"),
                ["properties"] = properties ?? new Dictionary<string, string>()
            };
            
            eventQueue.Enqueue(eventData);
            
            if (eventQueue.Count >= BatchSize)
            {
                StartCoroutine(FlushEvents());
            }
        }

        public void StartSession()
        {
            CheckInitialized();
            string newSessionId = System.Guid.NewGuid().ToString();
            this.userConfig = UserConfig.CreateBuilder()
                .SetSessionId(newSessionId)
                .SetUserId(userConfig.GetUserId())
                .SetAnonymous(userConfig.GetAnonymousId())
                .Build();
            
            SystemEvent()
                .Category(typeof(UserEvents))
                .Type(UserEvents.SESSION_START)
                .SetProperties(new Dictionary<string, string>())
                .Trigger();
        }

        public void EndSession()
        {
            CheckInitialized();
            
            if (string.IsNullOrEmpty(userConfig.GetSessionId()))
            {
                return;
            }
            
            SystemEvent()
                .Category(typeof(UserEvents))
                .Type(UserEvents.SESSION_END)
                .SetProperties(new Dictionary<string, string>())
                .Trigger();
        }

        public void UpdateUserConfig(UserConfig newUserConfig)
        {
            CheckInitialized();
            this.userConfig = newUserConfig;
        }
        
        public UserConfig GetUserConfig()
        {
            return userConfig;
        }
        
        public DeviceInfo GetDeviceInfo()
        {
            return deviceInfo;
        }

        private DeviceInfo AutoFetchDeviceInfo()
        {
            try
            {
                return new DeviceInfo(
                    GetPlatformString(),
                    SystemInfo.operatingSystem,
                    Application.version,
                    SystemInfo.deviceModel,
                    $"{Screen.width}x{Screen.height}",
                    SystemInfo.deviceName
                );
            }
            catch
            {
                return new DeviceInfo(
                    "UNITY",
                    "Unknown",
                    "1.0.0",
                    "Unknown",
                    "Unknown",
                    "Unknown"
                );
            }
        }

        private string GetPlatformString()
        {
            switch (Application.platform)
            {
                case RuntimePlatform.Android:
                    return "ANDROID";
                case RuntimePlatform.IPhonePlayer:
                    return "IOS";
                case RuntimePlatform.WindowsPlayer:
                case RuntimePlatform.WindowsEditor:
                    return "WINDOWS";
                case RuntimePlatform.OSXPlayer:
                case RuntimePlatform.OSXEditor:
                    return "MACOS";
                case RuntimePlatform.LinuxPlayer:
                    return "LINUX";
                case RuntimePlatform.WebGLPlayer:
                    return "WEBGL";
                case RuntimePlatform.XboxOne:
                    return "XBOX";
                case RuntimePlatform.PS4:
                case RuntimePlatform.PS5:
                    return "PLAYSTATION";
                case RuntimePlatform.Switch:
                    return "NINTENDO_SWITCH";
                default:
                    return "UNITY";
            }
        }

        private IEnumerator FlushEvents(bool force = false)
        {
            if (isSending || eventQueue.Count == 0) yield break;
            
            isSending = true;
            
            int count = Mathf.Min(BatchSize, eventQueue.Count);
            var batch = new List<Dictionary<string, object>>();
            
            for (int i = 0; i < count; i++)
            {
                batch.Add(eventQueue.Dequeue());
            }
            
            var json = JsonUtility.ToJson(new { events = batch });
            
            using (var request = new UnityWebRequest(baseUrl, "POST"))
            {
                byte[] bodyRaw = Encoding.UTF8.GetBytes(json);
                request.uploadHandler = new UploadHandlerRaw(bodyRaw);
                request.downloadHandler = new DownloadHandlerBuffer();
                request.SetRequestHeader("Content-Type", "application/json");
                request.SetRequestHeader("x-api-key", apiKey);
                
                yield return request.SendWebRequest();
            }
            
            isSending = false;
            
            if (force && eventQueue.Count > 0)
            {
                yield return FlushEvents(true);
            }
        }

        private void OnApplicationPause(bool pauseStatus)
        {
            if (pauseStatus)
            {
                StartCoroutine(FlushEvents());
            }
        }

        private void OnApplicationQuit()
        {
            if (isInitialized)
            {
                StartCoroutine(FlushEvents(true));
            }
        }

        private void CheckInitialized()
        {
            if (!isInitialized)
            {
                throw new InvalidOperationException("GameAlytics must be initialized first. Call GameAlytics.Init()");
            }
        }
    }
}

File: research/core/src/GameAlytics.ts
import {
  EventPayload,
  EventBuilder,
  EventProperties,
  GamepulseInterface,
  GamepulseConfig,
  EventType,
  EventCategory,
  UserEvents,
  GameplayEvents,
  EconomyEvents,
  ProgressionEvents,
  AdEvents,
  IAPEvents
} from './types';

class EventBuilderImpl implements EventBuilder {
  private properties: EventProperties = {};
  private readonly eventType: string;
  private readonly eventCategory: string;
  private readonly isCustom: boolean;

  constructor(eventType: string, eventCategory: string, isCustom: boolean) {
    this.eventType = eventType;
    this.eventCategory = eventCategory;
    this.isCustom = isCustom;
  }

  setProperties(properties: EventProperties): this {
    this.properties = { ...this.properties, ...properties };
    return this;
  }

  async track(): Promise<void> {
    await (Gamepulse.getInstance() as any).trackEvent({
      type: this.isCustom ? 'CUSTOM' : 'SYSTEM',
      value: this.eventType,
      category: this.eventCategory,
      properties: this.properties,
    });
  }
}

export class Gamepulse implements GamepulseInterface {
  private static instance: Gamepulse;
  private config!: GamepulseConfig;
  private isInitialized = false;
  private userId: string | null = null;
  private sessionId: string | null = null;
  private anonymousId: string;

  private constructor() {
    // Generate a unique anonymous ID
    this.anonymousId = this.generateUUID();
  }

  public static getInstance(): Gamepulse {
    if (!Gamepulse.instance) {
      Gamepulse.instance = new Gamepulse();
    }
    return Gamepulse.instance;
  }

  public static init(config: GamepulseConfig): Gamepulse {
    const instance = Gamepulse.getInstance();
    instance.initialize(config);
    return instance;
  }

  private initialize(config: GamepulseConfig): void {
    if (this.isInitialized) {
      console.warn('Gamepulse is already initialized');
      return;
    }

    if (!config.apiKey) {
      throw new Error('API key is required');
    }

    this.config = {
      debug: false,
      ...config,
    };

    this.isInitialized = true;
    
    if (this.config.debug) {
      console.log('Gamepulse initialized with config:', this.config);
    }
  }

  public event(type: string, category: string): EventBuilder {
    this.ensureInitialized();
    return new EventBuilderImpl(type, category, false);
  }

  public customEvent(type: string, category: string): EventBuilder {
    this.ensureInitialized();
    return new EventBuilderImpl(type, category, true);
  }

  public async startSession(): Promise<void> {
    this.ensureInitialized();
    this.sessionId = this.generateUUID();
    
    await this.event(UserEvents.SESSION_START, 'user')
      .setProperties({})
      .track();
  }

  public async endSession(): Promise<void> {
    this.ensureInitialized();
    
    if (!this.sessionId) {
      return;
    }
    
    await this.event(UserEvents.SESSION_END, 'user')
      .setProperties({})
      .track();
      
    this.sessionId = null;
  }

  public identify(userId: string): void {
    this.ensureInitialized();
    this.userId = userId;
  }

  public userEvent(eventType: string): EventBuilder {
    this.ensureInitialized();
    return new EventBuilderImpl(eventType, 'user', false);
  }

  public gameplayEvent(eventType: string): EventBuilder {
    this.ensureInitialized();
    return new EventBuilderImpl(eventType, 'gameplay', false);
  }

  public economyEvent(eventType: string): EventBuilder {
    this.ensureInitialized();
    return new EventBuilderImpl(eventType, 'economy', false);
  }

  public progressionEvent(eventType: string): EventBuilder {
    this.ensureInitialized();
    return new EventBuilderImpl(eventType, 'progression', false);
  }

  public adEvent(eventType: string): EventBuilder {
    this.ensureInitialized();
    return new EventBuilderImpl(eventType, 'ad', false);
  }

  public iapEvent(eventType: string): EventBuilder {
    this.ensureInitialized();
    return new EventBuilderImpl(eventType, 'iap', false);
  }

  public getDeviceInfo() {
    return {
      platform: this.getPlatform(),
      osVersion: this.getOSVersion(),
      deviceModel: this.getDeviceModel(),
      screenResolution: this.getScreenResolution(),
      deviceManufacturer: this.getDeviceManufacturer(),
      appVersion: this.config.appVersion || '1.0.0',
    };
  }

  public async trackEvent(payload: Omit<EventPayload, keyof ReturnType<Gamepulse['getDeviceInfo']> | 'timezone' | 'localDateTime' | 'userId' | 'anonymousId' | 'sessionId'>) {
    const deviceInfo = this.getDeviceInfo();
    const now = new Date();
    
    const eventPayload: EventPayload = {
      ...payload,
      ...deviceInfo,
      timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
      localDateTime: now.toISOString(),
      userId: this.userId || '',
      anonymousId: this.anonymousId,
      sessionId: this.sessionId || '',
    };

    if (this.config.debug) {
      console.log('Tracking event:', eventPayload);
    }

    try {
      const response = await fetch('https://client.dev.gamepulse.studio/events/collect', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': this.config.apiKey,
        },
        body: JSON.stringify(eventPayload),
      });

      if (!response.ok) {
        throw new Error(`Failed to track event: ${response.statusText}`);
      }
    } catch (error) {
      console.error('Error tracking event:', error);
      throw error;
    }
  }

  private ensureInitialized(): void {
    if (!this.isInitialized) {
      throw new Error('Gamepulse must be initialized first. Call Gamepulse.init()');
    }
  }

  // Platform-specific implementations (to be overridden by platform-specific classes)
  protected getPlatform(): string {
    return 'web';
  }

  protected getOSVersion(): string {
    return 'unknown';
  }

  protected getDeviceModel(): string {
    return 'unknown';
  }

  protected getScreenResolution(): string {
    if (typeof window !== 'undefined' && window.screen) {
      return `${window.screen.width}x${window.screen.height}`;
    }
    return 'unknown';
  }

  protected getDeviceManufacturer(): string {
    return 'unknown';
  }

  private generateUUID(): string {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === 'x' ? r : (r & 0x3) | 0x8;
      return v.toString(16);
    });
  }
}

// Export a singleton instance
export default Gamepulse.getInstance();

// Export types
export * from './types';
